diff -r fb76b171f510 lib/Sub/HandlesVia/Toolkit/Moose.pm
--- a/lib/Sub/HandlesVia/Toolkit/Moose.pm	Sat Jul 09 20:32:37 2022 +0100
+++ b/lib/Sub/HandlesVia/Toolkit/Moose.pm	Sat Jul 09 20:46:31 2022 +0100
@@ -15,9 +15,12 @@
 	my ($target) = @_;
 	
 	require Moose::Util;
-	my $meta = Moose::Util::find_meta($target);
-	Role::Tiny->apply_roles_to_object($meta, $me->package_trait);
-	Role::Tiny->apply_roles_to_object($meta, $me->role_trait) if $meta->isa('Moose::Meta::Role');
+	require Moose::Util::MetaRole;
+	Moose::Util::MetaRole::apply_metaroles(
+		for             => $target,
+		class_metaroles => { class => [ $me->package_trait ] },
+		role_metaroles  => { role  => [ $me->package_trait, $me->role_trait ] },
+	);
 }
 
 sub package_trait {
@@ -158,10 +161,10 @@
 our $AUTHORITY = 'cpan:TOBYINK';
 our $VERSION   = '0.031';
 
-use Role::Tiny;
+use Moose::Role;
 
 sub _shv_toolkit {
-	'Sub::HandlesVia::Toolkit::Moose',
+	'Sub::HandlesVia::Toolkit::Moose';
 }
 
 around add_attribute => sub {
@@ -196,24 +199,18 @@
 our $AUTHORITY = 'cpan:TOBYINK';
 our $VERSION   = '0.031';
 
-use Role::Tiny;
+use Moose::Role;
 
 around apply => sub {
 	my ($next, $self, $other, %args) = (shift, shift, @_);
-	
-	if ($other->isa('Moose::Meta::Role')) {
-		Role::Tiny->apply_roles_to_object(
-			$other,
-			$self->_shv_toolkit->package_trait,
-			$self->_shv_toolkit->role_trait,
-		);
-	}
-	else {
-		Role::Tiny->apply_roles_to_object(
-			$other,
-			$self->_shv_toolkit->package_trait,
-		);
-	}
+
+	require Moose::Util;
+	require Moose::Util::MetaRole;
+	Moose::Util::MetaRole::apply_metaroles(
+		for             => $other,
+		class_metaroles => { class => [ $self->_shv_toolkit->package_trait ] },
+		role_metaroles  => { role  => [ $self->_shv_toolkit->package_trait, $self->_shv_toolkit->role_trait ] },
+	);
 	
 	$self->$next(@_);
 };
